Issue#237

PureKrome opened the pull request:
Here we go :)

This Pull Requests is for the following added functionality:
- /Note <optional text>
- /Afk <optional text>

EDIT: I've also done a merge with upstream/master .. merge 2x conflicts .. committed the Merge .. and now Re-Pull Requested.

PureKrome committed:
- ChatUser now has a Note property.
- Implemented a 'Note' command for users. This includes initial code for
  setting/clearing the CharUser.Note property and some basic CSS/JQuery.
- Added first 2 unit tests, which are passing. More to come.
- Migration doesn't exist -> I'm trying to figure this out *blush*.

PureKrome committed:
- Removed and readded package EntityFramework.Migrations v 0.7.0.1 because
  there was no Add-Migrations PS command available.
- Add-Migration UserNote applied.

PureKrome committed:
- UI tweaks to correctly show a stock image if a ChatUser has a note.

PureKrome committed:
- Fixed compile time error with my Note unit tests. *blush*

PureKrome committed:
- Added Afk command and unit tests.

PureKrome committed:
- Replaced the Note image with Css.
- Css for notes now has two images :: one for notes, one for Afk.
- Typing now removes any Afk note image and state.

PureKrome committed:
- Fixed up the New User xml template so when a user is now added to a room,
  their Note image and tooltip will now render.

PureKrome committed:
- Added new Note (Paperclip) and Afk (StopWatch) png icons.
- Fixed up UI positioning for Note/Afk icons.

PureKrome committed:
- Code cleanup.

PureKrome committed:
- Added Afk command to the GetCommands list. (How did that get missed?!)

PureKrome committed:
Merge remote-tracking branch 'upstream/master' into Issue#237

Conflicts:
	JabbR/Hubs/Chat.cs
	JabbR/ViewModels/UserViewModel.cs

PureKrome committed:
- Afk text now displays correctly to a (valid) room.

davidfowl commented on JabbR/Chat.js:
```diff
@@ -393,6 +397,20 @@
+        ui.changeNote(user, room);
+
+        if (!isSelf(user)) {
+            ui.addMessage(user.Name + " has " + (user.IsAfk ? "gone Afk" : (user.Note == null ? "cleared" : "set") + " their note"), 'notification', room);
```
don't use double quotes, everywhere else uses single quotes.

Also clean this line up, way too much expressions in here, makes it unreadable

PureKrome replied:
Is this considered, cleaned up / easier to read?

``` javascript
// Called when you have added or cleared a note
chat.noteChanged = function (isAfk, isCleared) {
    var afkMessage = 'You have gone Afk';
    var noteMessage = 'Your note has been ' + isCleared ? 'cleared' : 'set';
    ui.addMessage(isAfk ? afkMessage: noteMessage, 'notification', this.activeRoom);
};
```

davidfowl replied:
Yup.

davidfowl commented on JabbR/Chat.ui.js:
```diff
@@ -1060,9 +1066,25 @@
     };
 
+    function toggleNote(user, $user) {
+        var element = $user.find('.note');
```
We use $element for references jQuery elements. In this case I'd call the variable $note.

PureKrome replied:
Awesome - done.

davidfowl commented on JabbR/Commands/CommandManager.cs:
```diff
@@ -766,5 +779,25 @@ private void HandleNudge(ChatRoom room, ChatUser user, string[] parts)
+
+        private void HandleNote(ChatUser user, string[] parts)
+        {
+            // NOTE (see what i did there!): "Parts" with only 1 item means we're clearing the note.
```
What is this comment?

Can you make this code clearer.

PureKrome replied:
Easier?

``` javascript
 private void HandleNote(ChatUser user, string[] parts)
        {
            // We need to determine if we're either
            // 1. Setting a new Note.
            // 2. Clearing the existing Note.
            // If we have no optional text, then we need to clear it. Otherwise, we're storing it.
            bool isNoteBeingCleared = parts.Length == 1;
            user.Note = isNoteBeingCleared ? null : String.Join(" ", parts.Skip(1)).Trim();

            _notificationService.ChangeNote(user);

            _repository.CommitChanges();
        }
```

davidfowl replied:
Might be simpler to not call HandleNode from HandleAfk.

PureKrome replied:
But what about DRY?

Afk is a note with some prepended text...

davidfowl replied:
I'll sacrifice DRY any day for clarity. Plus it's only 3 lines.

PureKrome replied:
here's the Afk code now. Yep, nicer.

``` javascript
private void HandleAfk(ChatUser user, string[] parts)
        {
            string message = String.Join(" ", parts.Skip(1)).Trim();
            user.Note = string.Format("{0} {1}", ChatUser.AfkPrependingText, message).Trim();

            _notificationService.ChangeNote(user);

            _repository.CommitChanges();
        }
```

davidfowl replied:
Uppercase S for string.Format.

davidfowl commented on JabbR/Hubs/Chat.cs:
```diff
@@ -682,6 +684,25 @@ void INotificationService.OnUserNameChanged(ChatUser user, string oldUserName, s
+            // Update the calling client
+            foreach (var client in user.ConnectedClients)
+            {
+                Clients[client.Id].noteChanged(user.IsAfk, user.Note == null);
```
calculate `user.Note == null` outside of the loop and give it a proper variable name.

PureKrome replied:
Good pickup. Fix0rd.

davidfowl commented on JabbR/Models/ChatUser.cs:
```diff
@@ -18,6 +18,13 @@ public class ChatUser
+        [NotMapped]
+        public bool IsAfk
+        {
+            get { return Note != null && Note.StartsWith("Afk"); }
```
Ignore case for the starts with? Maybe use a constant?

davidfowl commented on JabbR/ViewModels/UserViewModel.cs:
```diff
@@ -10,11 +10,16 @@ public UserViewModel(ChatUser user)
         public bool Active { get; set; }
         public string Status { get; set; }
+        public string Note { get; set; }
+        public string NoteCss { get { return Note == null ? null : IsAfk ? "afk" : "message"; } }
```
Split this into multiple lines. Not sure about putting the css class here, can't this be done in JavaScript?

PureKrome replied:
It can.. but IMO the View (which i see the JS, is) should not have any logic in it.

davidfowl replied:
IMO it can, there's different layers of view models. Take a look at all the other crap we do in javascript.

PureKrome replied:
I can ... but this is where it is mainly used:

Index.html / mobile.html

```
<span class="note{{if noteClass}} ${noteClass}{{/if}}" {{if note}}title="${note}"{{/if}}></span>
```

that line could/will get pretty messy....

davidfowl replied:
False, there's is js that is run before the template ever gets rendered. Just duplicate the logic in the client side js not in the template.

davidfowl commented on JabbR/JabbR.csproj:
```diff
@@ -45,7 +45,7 @@
       <HintPath>..\packages\EntityFramework.4.2.0.0\lib\net40\EntityFramework.dll</HintPath>
     </Reference>
-    <Reference Include="EntityFramework.Migrations">
+    <Reference Include="EntityFramework.Migrations, Version=0.7.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089, processorArchitecture=MSIL">
```
Did you upgrade?

PureKrome replied:
Nope. manually fixed the csproj.

PureKrome committed:
- Cleaned up chat.noteChanged and chat.changeNote code to make it more readable.
- Added // Assert comments :) !!! I usually never forget those !!!

PureKrome committed:
- Renamed var element to var $note for better readability.

davidfowl commented on JabbR/Chat.ui.js:
```diff
@@ -1060,9 +1066,25 @@
+        var $note = $user.find('.note');
+        $note.removeClass('afk message');
+        $note.removeAttr('title');
+        if (user.Note != null) {
```
`!==`

PureKrome replied:
Fixed. I've never seen that operator before. Learnt something new this evening :)

davidfowl commented on JabbR/Commands/CommandManager.cs:
```diff
@@ -766,5 +779,25 @@ private void HandleNudge(ChatRoom room, ChatUser user, string[] parts)
+        {
+            // First part is the command.
+            // Second is the AFK initial/prepending text.
+            var newParts = new List<string> {"afk", "Afk"};
```
Seems a little hacky

PureKrome committed:
- Removed NoteCss property from UserViewModel and replaced the logic into the
  appropriate javascript functions chat.getRoomInfo and chat.addUser.
- Fixed improper use of == with an === or !=== in functions toggleNote and chat.addUser.
- Cleaned up comments in HandleNote(..) method to better explain my logic.
- Refactored HandleAfk to now not call HandleNote(..).
- Refactored the bool IsAfk property for ChatUser.

PureKrome committed:
- Refactored the logic for determining the Note css class. It's now more DRY.
